### PR TITLE
feat(plat): added support to Nuvoton npcm845x chip

### PR DESCRIPTION
**Pull request by rutigl**
feat(plat): added support to Nuvoton npcm845x chip

Initial version

Signed-off-by: Margarita Glushkin [rutigl@gmail.com](mailto:rutigl@gmail.com)

### CODE REVIEW

File: core/arch/arm/plat-nuvoton/conf.mk [hunk 1]
```diff
@@ -0,0 +1,59 @@
+PLATFORM_FLAVOR ?= npcm845x
+
+$(info PLATFORM_FLAVOR is [${PLATFORM_FLAVOR}])
```
Comment (jforissier) [inline core/arch/arm/plat-nuvoton/conf.mk:3]: No

File: core/arch/arm/plat-nuvoton/conf.mk [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+ifeq ($(PLATFORM_FLAVOR),npcm845x)
+include core/arch/arm/cpu/cortex-armv8-0.mk
+CFG_ARM64_core = y
+endif #npcm845x
+CFG_USER_TA_TARGETS ?= ta_arm64
+$(call force,DEBUG,1)
```
Comment (jforissier) [inline core/arch/arm/plat-nuvoton/conf.mk:10]: `force` is not appropriate here. Use `DEBUG ?= 1` if you want it by default.

File: core/arch/arm/plat-nuvoton/conf.mk [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+$(call force,CFG_WITH_LPAE,y)
+else
+$(call force,CFG_ARM32_core,y)
+endif
+
+CFG_WITH_STATS ?= y
+
+
+ifeq ($(PLATFORM_FLAVOR),npcm845x)
+CFG_TEE_CORE_NB_CORE = 4
```
Comment (jforissier) [inline core/arch/arm/plat-nuvoton/conf.mk:46]: $(call force ...)

File: core/arch/arm/plat-nuvoton/main.c [hunk 4]
```diff
@@ -0,0 +1,188 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2016-2020, Linaro Limited
```
Comment (jforissier) [inline core/arch/arm/plat-nuvoton/main.c:3]: + Nuvoton copyright?

File: core/arch/arm/plat-nuvoton/platform_config.h [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+/* Make stacks aligned to data cache line length */
+#define STACK_ALIGNMENT		64
+
+#ifdef CFG_WITH_PAGER
+#error "Pager not supported for NPCM845x"
+#endif
+
+
+
```
Comment (jforissier) [inline core/arch/arm/plat-nuvoton/platform_config.h:20]: Too many blank lines

File: core/arch/arm/plat-nuvoton/platform_config.h [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+#define GIC_BASE		0xDFFF8000
+#define UART0_BASE		0xf0000000
+#define UART1_BASE		0xf0001000
+#define UART2_BASE		0xf0002000
+#define UART3_BASE		0xf0003000
+//#define TZC400_BASE		0x2a4a0000
```
Comment (jforissier) [inline core/arch/arm/plat-nuvoton/platform_config.h:33]: No commented out code please

File: core/arch/arm/plat-nuvoton/utils/npcm845x_trace.c [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+		boffs = sizeof(buf) - 2;
+
+	buf[boffs] = '\n';
+	while (boffs && buf[boffs] == '\n')
+		boffs--;
+	boffs++;
+	buf[boffs + 1] = '\0';
+
+	trace_ext_puts(buf);
+}
```
Comment (jforissier) [inline core/arch/arm/plat-nuvoton/utils/npcm845x_trace.c:63]: I am not sure this is needed at all, but anyways it should not be in a platform-specific file

File: core/arch/arm/plat-nuvoton/utils/npcm845x_utils.c [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+	TMSG(">================================================");
+	TMSG("OP-TEE OS Version %s", core_v_str);
+	TMSG("Nuvoton OP-TEE OS Version %d.%d.%d%s",
+			NPCMX845_OPTEE_VERSION_MAJOR,
+			NPCMX845_OPTEE_VERSION_MINOR,
+			NPCMX845_OPTEE_VERSION_PATCH,
+			NPCMX845_OPTEE_VERSION_BUILD);
+	TMSG(">================================================");
+	TMSG(COLOR_NORMAL);
+}
```
Comment (jforissier) [inline core/arch/arm/plat-nuvoton/utils/npcm845x_utils.c:40]: We certainly do not want any plaform-specific versioning scheme and neither patform-specific display code.

File: core/arch/arm/plat-nuvoton/conf.mk [hunk 9]
```diff
@@ -7,7 +7,6 @@ include core/arch/arm/cpu/cortex-armv8-0.mk
 CFG_ARM64_core = y
 endif #npcm845x
 CFG_USER_TA_TARGETS ?= ta_arm64
-$(call force,CFG_CORE_DYN_SHM,n)
```
Comment (jforissier) [inline core/arch/arm/plat-nuvoton/conf.mk]: Please do not add this line in the first commit rather than removing it later...

**Review by jforissier — CHANGES_REQUESTED** [review]
Please rework this series to make it minimal (not so many commits are needed) and see below comments. Thanks.

File: core/arch/arm/plat-nuvoton/npcm845x_utils.h [hunk 10]
```diff
@@ -0,0 +1,37 @@
+/*
+ *  Copyright (c) 2022 Nuvoton Technology Corp.
+ *
+ * See file CREDITS for list of people who contributed to this
+ * project.
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of
+ * the License, or (at your option) any later version.
```
Comment (etienne-lms) [inline core/arch/arm/plat-nuvoton/npcm845x_utils.h:10]: OP-TEE OS will not merge GPL licensed code. Please reconsider you licensing model.

File: core/arch/arm/plat-nuvoton/main.c [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+	tzc_init((vaddr_t)va);
+	tzc_dump_state();
+
+	return TEE_SUCCESS;
+}
+
+service_init(init_tzc400);
+#endif /*CFG_TZC400*/
+
+#if defined(PLATFORM_FLAVOR_qemu_virt)
```
Comment (etienne-lms) [inline core/arch/arm/plat-nuvoton/main.c:144]: ???

File: core/arch/arm/plat-nuvoton/main.c (hunk 4 above)
Comment (etienne-lms) [reply core/arch/arm/plat-nuvoton/main.c:3]: I think both copyright could apply

**Review by etienne-lms — COMMENTED** [review]
please run checkpatch script and set an initialization value to each local variable definition.

**Comment by rutigl** [comment]
You are right, I just wanted to fix the code according the reviewers messages.

2023-05-09 16:45 GMT+03:00, Jérôme Forissier ***@***.***>:

**Comment by jforissier** [comment]
> You are right, I just wanted to fix the code according the reviewers messages. 

That's good, but it should be obvious which commit you are fixing (for example by appending a `[Review]` or `[Fixup]` prefix, see https://optee.readthedocs.io/en/latest/general/contribute.html#finalizing-your-contribution) and there should be no merge commits.
Before pushing you can use `git log` and `git show <sha1>` to review each commit and ask yourself if they all make sense for a reviewer to see.

File: core/arch/arm/plat-nuvoton/conf.mk [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+endif
+
+ifeq ($(platform-flavor-armv8),1)
+CFG_WITH_ARM_TRUSTED_FW = y
+endif
+
+CFG_GIC = y
+CFG_ARM_GICV2 = y
+CFG_SECURE_TIME_SOURCE_CNTPCT = y
+CFG_16550_UART = y
```
Comment (etienne-lms) [inline core/arch/arm/plat-nuvoton/conf.mk:23]: Likely `$(call force,...)` here

File: core/arch/arm/plat-nuvoton/main.c [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+register_ddr(DRAM1_BASE, DRAM1_SIZE);
+#endif /* DRAM1_BASE */
+
+#ifdef GIC_BASE
+register_phys_mem_pgdir(MEM_AREA_IO_SEC, GICD_BASE, GIC_DIST_REG_SIZE);
+register_phys_mem_pgdir(MEM_AREA_IO_SEC, GICC_BASE, GIC_DIST_REG_SIZE);
+
+void main_init_gic(void)
+{
+print_version();
```
Comment (etienne-lms) [inline core/arch/arm/plat-nuvoton/main.c:51]: Indentation. Please run checkpatch on your changes. 
See [checkpatch documentation](https://optee.readthedocs.io/en/latest/general/coding_standards.html?highlight=checkpatch#running-checkpatch) as well as `scripts/checkpatch.pl --help`.

File: core/arch/arm/plat-nuvoton/main.c [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+#if defined(CFG_WITH_ARM_TRUSTED_FW)
```
Comment (etienne-lms) [inline core/arch/arm/plat-nuvoton/main.c:52]: Do you plan to support both config: `CFG_WITH_ARM_TRUSTED_FW=y` as well as `CFG_WITH_ARM_TRUSTED_FW=n`?
As your platform is 64bit capable, I strongly think you will always rely on `CFG_WITH_ARM_TRUSTED_FW=y`. In that case use `$(call force,CFG_WITH_ARM_TRUSTED_FW,y)` in _conf.mk_ and remove cases where `CFG_WITH_ARM_TRUSTED_FW` is not defined.

File: core/arch/arm/plat-nuvoton/main.c [hunk 15]
```diff
@@ -0,0 +1,81 @@
+/*
```
Comment (etienne-lms) [inline core/arch/arm/plat-nuvoton/main.c:1]: OP-TEE requires SPDX tags, see [documentation](https://optee.readthedocs.io/en/latest/general/license_headers.html?highlight=SPDX#license-headers).

File: core/arch/arm/plat-nuvoton/main.c [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+#include <stdint.h>
+#include <string.h>
+#include <trace.h>
+
+static struct gic_data gic_data __nex_bss;
+static struct ns16550_data console_data __nex_bss;
+
+register_phys_mem_pgdir(MEM_AREA_IO_SEC, CONSOLE_UART_BASE, UART_REG_SIZE);
+
+#ifdef DRAM0_BASE
```
Comment (etienne-lms) [inline core/arch/arm/plat-nuvoton/main.c:37]: check your _platform_config.h_: DRAM0_BASE is always defined and DRAM1_BASE is never defined. Please remove uncessarry conditions/code.

File: core/arch/arm/plat-nuvoton/platform_config.h [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+#define UART2_BASE	0xf0002000
+#define UART3_BASE	0xf0003000
+#define UART_REG_SIZE	0x100
+#define CONSOLE_UART_BASE	UART0_BASE
+#define CONSOLE_UART_CLK_IN_HZ	1
+
+#define GICD_OFFSET	0x1000
+#define GICC_OFFSET	0x2000
+
+#ifdef GIC_BASE
```
Comment (etienne-lms) [inline core/arch/arm/plat-nuvoton/platform_config.h:45]: of course it is defined, right above.

File: core/arch/arm/plat-nuvoton/utils/npcm845x_trace.c [hunk 18]
```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright (c) 2022-2023, ARM Limited and Contributors. All rights reserved.
+ *
+ * Copyright (C) 2022-2023 Nuvoton Ltd.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#if defined(__KERNEL__)
```
Comment (etienne-lms) [inline core/arch/arm/plat-nuvoton/utils/npcm845x_trace.c:9]: not needed

File: core/arch/arm/plat-nuvoton/utils/npcm845x_trace.c [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+#include <platform_config.h>
+#endif
+
+#include <utils/npcm845x_trace.h>
```
Comment (etienne-lms) [inline core/arch/arm/plat-nuvoton/utils/npcm845x_trace.c:13]: sort all header files inclusions in alphabetical order.

File: core/arch/arm/plat-nuvoton/conf.mk [hunk 20]
```diff
@@ -0,0 +1,54 @@
...
+CFG_TEE_CORE_NB_CORE = 4
+# [3000.0000 031f.ffff] is reserved to early boot
+CFG_TZDRAM_START ?= 0x36000000
+CFG_TZDRAM_SIZE  ?= 0x00f00000
+# SHM chosen arbitrary, in a way that it does not interfere
+# with initial location of linux kernel, dtb and initrd.
+CFG_SHMEM_START ?= 0x2FD00000 #0x42000000   # Hila to check 
+CFG_SHMEM_SIZE  ?= 0x00200000
+# When Secure Data Path is enable, last MByte of TZDRAM is SDP test memory.
+CFG_TEE_SDP_MEM_SIZE ?= 0x00400000
```
Comment (etienne-lms) [inline core/arch/arm/plat-nuvoton/conf.mk:51]: Are you shure you want to support SDP?
At this stage of your platform port, I don't think you are.

File: core/arch/arm/plat-nuvoton/utils/npcm845x_utils.c (hunk 8 above)
Comment (etienne-lms) [reply core/arch/arm/plat-nuvoton/utils/npcm845x_utils.c:20]: `boot_init_primary_late()` already does the job, yet not in magenta.
Maybe you could prpose a generic patch to enable per-trace-level colorzation on OP-TEE trace message? That could be nice.

**Review by etienne-lms — COMMENTED** [review]
Please squash all your commits into a single one and run checkpatch over it.
Please also don't use git merge points in your patch series, at least in the patch series you create OP-TEE pull requests on.

**Comment by etienne-lms** [comment]
@rutigl & @jforissier, sorry, I had some pending comment and did not see you were discussing on the serie when I posted them. Sorry for the unnecessary flood.

**Comment by jenswi-linaro** [comment]
Let me know once you've removed all merge commits and fixed up the remaining commits properly and I'll take a look.

**Pull request closed without merging**